RuntimeLib/StellarMastery.c

```diff
@@ -0,0 +1,166 @@
+#include "StellarMastery.h"
+#include "Character.h"
+#include "Constants.h"
+#include "Runtime.h"
+#include "Force.h"
+
+
+/*
+Rolls single option from the pool.
+
+params:
+- __POOL__ - pool to roll from.
+- __COUNT__ - number of options in the pool.
+- __GET_CHANCE__ - (*PoolItemType) -> Int32 - callback to get the chance of the option.
+- __RESULT_REF__ - pointer to the result.
+*/
+#define ROLL_POOL(__POOL__, __COUNT__, __GET_CHANCE__, __RESULT_REF__)    \
+    {                                                                     \
+        Int32 MaxRandomValue = 0;                                         \
+        for (Int32 Index = 0; Index < __COUNT__; Index += 1) {            \
+            MaxRandomValue += __GET_CHANCE__(&__POOL__[Index]);           \
+        }                                                                 \
+                                                                          \
+        Int32 Seed = (Int32)PlatformGetTickCount();                       \
+        Int32 RandomValue = RandomRange(&Seed, 0, MaxRandomValue);        \
+        Int32 CumulativeChance = 0;                                       \
+                                                                          \
+        for (Int32 Index = 0; Index < __COUNT__; Index += 1) {            \
+            CumulativeChance += __GET_CHANCE__(&__POOL__[Index]);         \
+            if (RandomValue < CumulativeChance) {                         \
+                __RESULT_REF__ = &__POOL__[Index];                        \
+                break;                                                    \
+            }                                                             \
+        }                                                                 \
+    }
+
+
+
+RTDataStellarLineGradeRef RTDataStellarMasteryGetLineGrade(
+	RTRuntimeRef Runtime,
+	RTDataStellarLineRef StellarLine,
+	UInt32 MaterialCount
+) {
+	RTDataStellarLineGradeRef OutGrade = NULL;
+	
+	for (UInt32 i = 0; i < StellarLine->StellarLineGradeCount; i++) {
+		RTDataStellarLineGradeRef Grade = &StellarLine->StellarLineGradeList[i];
+		if (Grade->ItemCount == MaterialCount) {
+			OutGrade = Grade;
+			break;
+		}
+	}
+	return OutGrade;
+}
+
+
+RTStellarMasterySlotRef RTCharacterStellarMasteryGetSlot(
+	RTCharacterRef Character,
+	UInt8 GroupID,
+	UInt8 SlotLine,
+	UInt8 SlotIndex
+) {
+	RTStellarMasterySlotRef OutSlot = NULL;
+
+	for (UInt32 i = 0; i < Character->Data.StellarMasteryInfo.Info.SlotCount; i++) {
+		RTStellarMasterySlotRef Slot = &Character->Data.StellarMasteryInfo.Slots[i];
+		if (Slot->GroupID == GroupID && Slot->SlotLine == SlotLine && Slot->SlotIndex == SlotIndex) {
+			OutSlot = Slot;
+			break;
+		}
+	}
+	return OutSlot;
+}
+
+RTStellarMasterySlotRef RTCharacterStellarMasteryAssertSlot(
```

Comment:
What does AssertSlot mean?

Reply:
Function creates new or returns existing mastery slot from character's data, so "assert" is closest word I know to that operation, like "ensure that something exists"